### DISCUSSION TITLE
Help Needed for Production Reverse Proxy Setup

### DISCUSSION THREAD

**Discussion by baba-dev** [open-webui/open-webui#1235] [General]
# THIS IS NOT OFFICIALLY SUPPORTED BY OPEN WEBUI.

Hello, I have searched the forums, Issues, Reddit and Official Documentations for any information on how to reverse-proxy Open WebUI via Nginx.

Any assistance would be greatly appreciated. It seems there are JS errors when you force rewrite domain name via nginx. 

![image](https://github.com/open-webui/open-webui/assets/60427476/81e5829f-be4e-4fc8-892b-3331e1e2d49e)

**Comment by baba-dev** [comment]
Here is my existing Nginx Configuration:

        location /chatgpt {
            sub_filter '"/_app/' '"/chatgpt/_app/'; # Adding a rewrite rule for /_app/ based links.
            sub_filter '"/themes/' '"/chatgpt/themes/'; # Similar rewrite rule for themes based on console errors.
            sub_filter_once off;
            rewrite /chatgpt/(.*) /$1 break;
            proxy_pass http://192.168.173.252:8080;
        }

        location /chatgpt/api {
            rewrite /chatgpt/(.*) /$1 break;
            proxy_buffering off;
            proxy_set_header Origin '';
            proxy_set_header Referer '';
            proxy_pass http://192.168.173.252:8080/api;
        }

        location /chatgpt/api/v1 {
            rewrite /chatgpt/(.*) /$1 break;
            proxy_pass http://192.168.173.252:8080/api/v1;
        }

        location /chatgpt/themes {
            rewrite /chatgpt/(.*) /$1 break;
            proxy_pass http://192.168.173.252:8080/themes;
        }

**Reply by brtptrs** [reply]
Has anyone set this up successfully using apache?

**Comment by baba-dev** [comment]
Works on its Subdomain using the below config:

    server {
       listen 80;
       listen [::]:80;
       server_name chatgpt.local;

    location / {
      proxy_pass http://192.168.173.252:8080;
    }

    location /ollama/api {
      proxy_buffering off;
      proxy_set_header Origin '';
      proxy_set_header Referer '';
      proxy_pass http://192.168.173.252:8080/api;

    }

    location /api/v1 {
     proxy_pass http://192.168.173.252:8080/api/v1;
    }
    }

**Comment by justinh-rahb** [comment]
Given the need to make the entire backend application accessible under `/chatgpt/` to the outside world without the backend being aware of this URL base path, you can simplify the NGINX configuration. The primary goal here is to prepend the `/chatgpt` path segment for external users, while internally directing requests as if they were made to the root path `/`.

Here's a streamlined approach to meet these requirements:

### Simplified NGINX Configuration

This configuration ensures all requests that start with `/chatgpt/` are internally redirected to the root path of the backend application. It avoids specifying multiple locations for different backend paths, treating all requests uniformly.

```nginx
location /chatgpt/ {
    # Rewrite the incoming request by removing the /chatgpt prefix
    rewrite ^/chatgpt/(.*)$ /$1 break;

    # Proxy the modified request to the backend application
    proxy_pass http://192.168.173.252:8080/;

    # Optional: Set headers that might be required by the backend or for compliance with CORS, etc.
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-Proto $scheme;

    # Handle WebSocket upgrades if your application uses WebSockets
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
}
```

### Key Points

- **Unified Handling**: This configuration doesn't differentiate between `/chatgpt/api`, `/chatgpt/api/v1`, or any other specific paths. Instead, it treats all requests starting with `/chatgpt/` in the same manner, rewriting them to remove the `/chatgpt` prefix before proxying to the backend. This means that a request to `http://yourserver/chatgpt/api/v1/resource` will be proxied to `http://192.168.173.252:8080/api/v1/resource`.

- **Simplified Configuration**: By not dividing the configuration into multiple specific locations for API paths, themes, etc., maintenance becomes simpler. This approach assumes that the backend application can handle all requests from the root path and that there's no need for external paths to differ in handling from the NGINX's perspective.

- **Backend Transparency**: The backend application receives requests as if they were made directly to its root path, with no knowledge of the `/chatgpt` prefix. This setup meets the requirement for the backend to be "none the wiser" about the URL path modifications made by NGINX.

- **WebSocket Support**: If your application uses WebSockets, the additional headers `Upgrade` and `Connection "upgrade"` are crucial for supporting WebSocket connections through NGINX.

This configuration approach should meet the requirement of making the backend application accessible under `/chatgpt/` from the internet, with a setup that's straightforward to manage and does not necessitate the backend to be aware of the external path prefix.

**Reply by BlackHeart-TF** [reply]
this doesnt work. open-webui generates links to /static/... which will 404 because its at /chatgpt/static/

**Reply by samdelong** [reply]
This response appears to be generated by ChatGPT or some other AI...

**Comment by baba-dev** [comment]
Thank you. So I tried your config for subpath serving. Still stuck on similar errors:

![image](https://github.com/open-webui/open-webui/assets/60427476/b7580983-001a-4d59-903a-2366f52878dd)

**Comment by justinh-rahb** [comment]
@baba-dev ruled out cache issue? `Ctrl + Shift + R` or disabling cache in dev tools should clear that.

**Comment by baba-dev** [comment]
Yes, Tried in incognito as well.

**Comment by baba-dev** [comment]
I Have noticed yet another issue, with its Subdomain specific configuration for Nginx, The downloaded models do not show up on the proxied frontend :

![image](https://github.com/open-webui/open-webui/assets/60427476/b0dffa42-3454-440a-a90b-5a2ce19e4c00)

The Domain Specific Configuration in reference to:

    server {
       listen 80;
       listen [::]:80;
       server_name chatgpt.local;

    location / {
      proxy_pass http://192.168.173.252:8080;
    }

    location /ollama/api {
      proxy_buffering off;
      proxy_set_header Origin '';
      proxy_set_header Referer '';
      proxy_pass http://192.168.173.252:8080/api;
    }

    location /api/v1 {
     proxy_pass http://192.168.173.252:8080/api/v1;
    }
    }

I do have models working and running when I access the default address:

Image 2:
![image](https://github.com/open-webui/open-webui/assets/60427476/d842cca4-48d0-491b-aef7-bca856a2a5a6)

**Comment by baba-dev** [comment]
A Hint to as why it may not be working when proxied, There are some JS errors in the browser window:

![image](https://github.com/open-webui/open-webui/assets/60427476/45f9e6bb-c9af-4688-a29f-e7ffa78d50a8)

The Server Connection Fails line seems to be the main cause. On a normal acess to IP, There are no errors in console window.

**Comment by lesca** [comment]
Hello, 
this is a very good discussion to make webui work behind https proxy.

I've tested in my environment, configuration as below:
```conf
    location /chatgpt/ {
        proxy_buffering off;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header Accept-Encoding "";
        sub_filter '"/_app/' '"/chatgpt/_app/';
        sub_filter '"/themes/' '"/chatgpt/themes/';
        sub_filter_once off;

        # proxy
        proxy_pass http://192.168.2.2:3000/;
    }
```

Screenshot below shows it works at first, but when it comes to "**version.json**" request, it fails. Look into the headers, it seems the request URL is wrong. I suspect the loaded JS code doesn't respect the custom path **"/chatgpt"**, instead, it just appends its path (_app) after the hostname (or IP). 

To be short:
* Wrong: `https://u.silkdiy.cn/_app/version.json`
* Correct: `https://u.silkdiy.cn/chatgpt/_app/version.json`

![image](https://github.com/open-webui/open-webui/assets/5976219/0bd6cf76-305d-403d-91d0-11ff80c45186)

Then, I turned to below configuration using the **root virtual directory ("/")** for proxing. This works without problem.

```config
    location / {
        proxy_buffering off;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header Accept-Encoding "";
        # sub_filter '"/_app/' '"/chatgpt/_app/';
        # sub_filter '"/themes/' '"/chatgpt/themes/';
        # sub_filter_once off;

        # proxy
        proxy_pass http://192.168.2.2:3000/;
    }
```

My guess is something needs to be changed in the JS code to make it work with customized virtual directory of reversed proxy.

**Reply by mylegitches** [reply]
Mine is constantly 'losing' the base uri.  if you edit the address in the element and insert the base uri, things begin to work, albeit not a solution; merely a clue?  subdomain works fine.  I didn't look, but it would make sense that there is already an FR for adding the ability to set a base URI.

**Reply by xXNaVeEXx** [reply]
hey thank you that works
you just forgot the websocket
```
location / {
          proxy_buffering off;
          proxy_set_header Host $host;
          proxy_set_header X-Real-IP $remote_addr;
          proxy_set_header Accept-Encoding "";
          proxy_http_version 1.1;
          proxy_set_header Upgrade $http_upgrade;
          proxy_set_header Connection "upgrade";
        # sub_filter '"/_app/' '"/openwebui/_app/';
        # sub_filter '"/themes/' '"/openwebui/themes/';
        # sub_filter_once off;

        # proxy
        proxy_pass http://192.168.2.2:3000/;
       }
```

**Reply by vmadhavaram** [reply]
This NGINX reverse proxy configuration covers all scenarios (except user.png), I discovered so far.
Added **sub_filter_types** to include javascript files.  

I wish  **openwebui** team start prioritizing **BASE_URL** work. This is a basic feature and most needed.

```
 location /myopenwebui {
        proxy_pass http://webui:8080;
        rewrite ^/myopenwebui/(.*)$ /$1 break;
        sub_filter_types text/plain text/html application/javascript;
        sub_filter 'href="/' 'href="/myopenwebui/';
        sub_filter 'src="/' 'src="/myopenwebui/';
        sub_filter '"/_app/' '"/myopenwebui/_app/';
        sub_filter '_app/' 'myopenwebui/_app/';
        sub_filter '"/_app' '"/myopenwebui/_app';
        sub_filter '"/themes/' '"/myopenwebui/themes/';
        sub_filter '"/themes' '"/myopenwebui/themes';
        sub_filter 'themes' 'myopenwebui/themes';
        sub_filter '"/openai/' '"/myopenwebui/openai/';
        sub_filter 'user.png' 'myopenwebui/user.png';
        sub_filter '"/favicon.png' '"/myopenwebui/favicon.png';
        sub_filter 'api/' 'myopenwebui/api/';
        sub_filter 'ollama/' 'myopenwebui/ollama/';
        sub_filter_once off;

        # Add WebSocket support
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";

        # Standard headers
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        # Timeouts for WebSocket connections
        proxy_read_timeout 60s;
        proxy_send_timeout 60s;
    }

    location /ollama {
        proxy_pass http://ollama:11434;
        rewrite ^/ollama/(.*)$ /$1 break;

        # Add WebSocket support
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";

        # Standard headers
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        # Timeouts for WebSocket connections
        proxy_read_timeout 60s;
        proxy_send_timeout 60s;
    } 
```

**Comment by ProjectMoon** [comment]
I"m trying to expose Open WebUI via nginx. It works fine for the UI and stuff, but the proxied ollama API has issues with event streaming. I've narrowed it down to nginx itself being the problem, and turned off proxy_buffering, but it still does not seem to chunk correctly. Any ideas?

```
    location /ollama/api {
        proxy_set_header Connection "";
        proxy_http_version 1.1;
        proxy_request_buffering off;
        chunked_transfer_encoding off;
        proxy_buffering off;
        proxy_max_temp_file_size 0;
        proxy_cache off;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_pass http://localhost:11435/ollama/api;
    }
```

**Reply by EmilSchurr** [reply]
I've had the same issue, my deployment being ollama installed directly on the VM and the Open WebUI installed as a docker container. What fixed it was changing
`location /ollama/api`
to
`location /api`

**Reply by ProjectMoon** [reply]
Could you share a bit more about this? Maybe the config block? Are you mapping `/api` to `/ollama/api` for example?

**Comment by mila76** [comment]
Hi, setup a reverse proxy (with Nginx Proxy Manager) for https WebUI access.. Work but not very well
All the chat dont "stream" (i have block of response) anymore or dont work at all

I can't figure out wich options or modification can made to work with the reverse proxy.
I'm not an expert of reverse proxy so any help is welcome. Thanks

**Reply by mila76** [reply]
For reference the only missing option (added on the advance tab) is proxy_buffering off;
Thanks for @EmilSchurr  for sharing

**Comment by sprklinginfo** [comment]
can't use root to proxy to the WebUI container as I have other applications running. anyone have a working NGINX reverse proxy from a sub-directory in this case? thanks!

**Comment by EmilSchurr** [comment]
This doesn't seem to be the problem from the OP, but if anyone is still having trouble with a reverse proxy and getting the streaming to work correctly, this is the right config:

```
server {
    server_name subdomain.yourdomain.com;
    location / {
        proxy_pass http://10.10.3.10:8080;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $host;
        proxy_buffering off;
        proxy_set_header Origin '';
        proxy_set_header Referer '';
    }
    location /api {
        proxy_http_version 1.1;
        proxy_buffering off;
        proxy_set_header Origin '';
        proxy_set_header Referer '';
        proxy_pass http://10.10.3.10:8080/api;
    }
    location /api/v1 {
        proxy_pass http://10.10.3.10:8080/api/v1;
    }
    listen 443 ssl;
}
```

**Reply by ProjectMoon** [reply]
Is this specifically for Open WebUI, or the ollama process itself? Open WebUI proxies ollama under a /ollama endpoint, unless I'm missing something?

**Reply by ProjectMoon** [reply]
So I have set this up, and unlike before, the streaming sometimes actually works correctly! But a lot of times, it still apparently fails to chunk correctly, at least for `aichat` (CLI AI client).

**Reply by ProjectMoon** [reply]
I added a PR to `aichat` to make it use a proper JSON streaming parser (which the developer had already built for the project), and now `aichat` and everything else I use streams properly through nginx..

**Comment by sprklinginfo** [comment]
thanks for sharing @EmilSchurr. I have Ollama installed on a ubuntu 22 VM directly and it can be accessed from "http://127.0.0.1:11434".  I installed the open webui with the following docker command: `$ docker run -d --network=host -v open-webui:/app/backend/data -e OLLAMA_BASE_URL=http://127.0.0.1:11434 --name open-webui --restart always ghcr.io/open-webui/open-webui:main`. 
On the VM I configured the reverse proxy like the following blocks by incorporating yours, wishing to access with a subdirectory as I have other applications running on the root of Nginx and some other applications too. 
```shell
server {
    listen 443 ssl;
    server_name subdomain.exmple.com;
    root /var/www/web-app/public;
    index index.html index.htm index.php

    location /chat {
        proxy_pass http://127.0.0.1:8080;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $host;
        proxy_buffering off;
        proxy_set_header Origin '';
        proxy_set_header Referer '';
    }

 location /api {
        proxy_http_version 1.1;
        proxy_buffering off;
        proxy_set_header Origin '';
        proxy_set_header Referer '';
        proxy_pass http://127.0.0.1:8080/api;
    }
    location /api/v1 {
        proxy_pass http://127.0.0.1:8080/api/v1;
    }

  location / {
        try_files $uri $uri/ /index.php?$query_string;
    }
 location ~ \.php$ {
        try_files $uri =404;
        fastcgi_pass unix:/run/php/php8.1-fpm.sock;
        fastcgi_index index.php;
        fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
        include fastcgi_params;
    }
    location ^~ /tools {
        alias /var/www/tools;
    }
}```. 
but got lots of 404 errors. 
![Snipaste_2024-05-07_14-27-56](https://github.com/open-webui/open-webui/assets/3521826/4684c0ed-c743-4fdc-821f-597e6f6cef47)

**Comment by leo9800** [comment]
I'd like to share my configuration for Open-WebUI and ollama in docker stack and Apache HTTP Server (`httpd`, `apache2`) as reverse proxy. This configuration (with sensitive fields redacted) is used for production.

----------------

#### `docker-compose.yml` (Docker)

```
services:
  ollama:
    image: ollama/ollama
    container_name: ollama
    volumes:
      - ./ollama:/root/.ollama
    tty: true
    restart: unless-stopped

  open-webui:
    image: ghcr.io/open-webui/open-webui
    container_name: open-webui
    volumes:
      - ./open-webui:/app/backend/data
    depends_on:
      - ollama
    ports:
      - 127.0.0.1:50100:8080
    environment:
      - OLLAMA_BASE_URL=http://ollama:11434
      - WEBUI_SECRET_KEY=REDACTED
      - GLOBAL_LOG_LEVEL=WARNING
    restart: unless-stopped
```

#### `open-webui.conf` (Apache)
```
<VirtualHost *:443>
	ServerName open-webui.domain.tld
	SSLEngine On
	SSLCertificateFile /path/to/cert.pem
	SSLCertificateKeyFile /path/to/key.pem

	AllowEncodedSlashes NoDecode

	ProxyPass / http://127.0.0.1:50100/ nocanon upgrade=websocket timeout=3600
	ProxyPreserveHost On

	RequestHeader set "X-Real-IP" expr=%{REMOTE_ADDR}
	RequestHeader set "X-Forwarded-For" expr=%{REMOTE_ADDR}
	RequestHeader set "X-Forwarded-Port" expr=%{SERVER_PORT}
	RequestHeader set "X-Forwarded-Proto" expr=%{REQUEST_SCHEME}
	RequestHeader set "X-Forwarded-Ssl" expr=%{HTTPS}
</VirtualHost>
```

**Comment by mattbrth** [comment]
Can someone give indications to configure this for NPM (Nginx Proxy Manager) ? Thank you very much !

**Reply by harddal** [reply]
Enabling websocket support with the latest version of webui has been working for me. 
![Screenshot 2025-01-12 204908](https://github.com/user-attachments/assets/5e0fa06e-e628-4afc-8128-b11648903b2f)

**Comment by BlackHeart-TF** [comment]
does anything here actually address the issue that webui generates links and is unaware of the nginx config? i cant pass the splash screen because its wanting to find `/static/splash.png` which is actually at `/chat/static/splash.jpg`

**Comment by lmaddox** [comment]
Anybody got a working nginx config? I want to serve open-webui from /webui. I'm trying the aformentioned rewrite rules to :3003/, but I don't think it can work unless there's a way to inform webui that it's behind the reverse proxy, so that it adjusts its generated URLs. Likely that's the cause of our 404s.

update 1: found it (WEBUI_URL) [here](https://docs.openwebui.com/getting-started/env-configuration). I've disabled the rewrite rule. nginx is still giving 504s and 499s

update 2: got frustrated and tried using a different vhost. no avail
```
  webui:
    environment:
      #- WEBUI_URL=https://InnovAnon-Inc.chickenkiller.com/webui/
      - WEBUI_URL=https://lmaddox.chickenkiller.com
      - PORT=3003
```

```
  server_name lmaddox.chickenkiller.com;
  ssl_certificate ...
  location / {
    proxy_pass http://...:3003/;
    ...
  }
```

Update 3: I think I've got it working now. Not the SSL reverse proxy, but the hardware access for which it was a means to an end. See [here](https://stackoverflow.com/questions/59595548/microphone-permissions-always-denied-in-firefox-for-android-in-my-environment-b). It'd be cool to be able to wrap this with SSL, though.

**Comment by liv-io** [comment]
I was able to get Open WebUI working behind a Apache2 reverse proxy. Actually, the application is hidden behind 2 reverse proxies. Apache2 as perimeter reverse proxy with ModSecurity and Nginx as application reverse proxy for the Open WebUI running inside of a Podman container.

Maybe it helps someone out there.

### Apache2:

```
<VirtualHost *:80>
  ServerName ollama.example.com

  ErrorLog ${APACHE_LOG_DIR}/ollama.example.com-80_error_log
  CustomLog ${APACHE_LOG_DIR}/ollama.example.com-80_access_log combined
  LogLevel warn

  Header always set X-Frame-Options 'deny'
  Header always set X-Content-Type-Options 'nosniff'
  Header always set X-XSS-Protection '1; mode=block'

  Redirect permanent /  https://ollama.example.com
</VirtualHost>

<VirtualHost *:443>
  ServerName ollama.example.com

  LogFormat "remote_addr:%h %l remote_user:%u %t request:\"%r\" status:%>s %b \"%{Referer}i\" \"%{User-agent}i\" %{SSL_PROTOCOL}x %{SSL_CIPHER}x" combinedssl
  ErrorLog ${APACHE_LOG_DIR}/ollama.example.com-443_error_log
  CustomLog ${APACHE_LOG_DIR}/ollama.example.com-443_access_log combinedssl
  LogLevel warn

  Header always set X-Frame-Options 'deny'
  Header always set X-Content-Type-Options 'nosniff'
  Header always set X-XSS-Protection '1; mode=block'

  SSLEngine On
  SSLCompression Off
  SSLProtocol -ALL +TLSv1.2 +TLSv1.3
  SSLHonorCipherOrder On
  SSLCipherSuite 'TLS_AES_256_GCM_SHA384:AES256+EECDH'
  SSLCertificateFile '/etc/letsencrypt/live/example.com/cert.pem'
  SSLCertificateKeyFile '/etc/letsencrypt/live/example.com/privkey.pem'
  SSLCertificateChainFile '/etc/letsencrypt/live/example.com/chain.pem'
  SSLUseStapling  On
  SSLOCSPProxyURL "http://proxy.example.com:3128"
  SSLProxyEngine On

  # ModSecurity
  SecRuleEngine On

  ProxyPass        /ws/socket.io/ ws://ollama.example.com/socket.io/
  ProxyPassReverse /ws/socket.io/ ws://ollama.example.com/socket.io/

  ProxyPass        / https://ollama.example.com/
  ProxyPassReverse / https://ollama.example.com/

</VirtualHost>
```

### Nginx

```
server {
    listen *:80;
    server_name ollama.example.com;

    error_log /var/log/nginx/ollama.example.com-80_error.log;
    access_log /var/log/nginx/ollama.example.com-80_access.log main;

    add_header Strict-Transport-Security 'max-age=31536000' always;
    add_header X-Frame-Options 'deny' always;
    add_header X-Content-Type-Options 'nosniff' always;
    add_header X-XSS-Protection '1; mode=block' always;

    return 301 https://$server_name$request_uri;
}

server {
    listen *:443 ssl;
    server_name ollama.example.com;

    error_log /var/log/nginx/ollama.example.com-443_error.log;
    access_log /var/log/nginx/ollama.example.com-443_access.log main_ssl;

    add_header Strict-Transport-Security 'max-age=31536000' always;
    add_header X-Frame-Options 'deny' always;
    add_header X-Content-Type-Options 'nosniff' always;
    add_header X-XSS-Protection '1; mode=block' always;

    ssl_protocols TLSv1.2 TLSv1.3;
    ssl_prefer_server_ciphers on;
    ssl_ciphers 'AES256+EECDH:AES256+EDH';
    ssl_ecdh_curve 'secp384r1';
    ssl_dhparam '/etc/nginx/dhparam.pem';
    ssl_certificate '/etc/ssl/example/ollama.example.com.crt';
    ssl_certificate_key '/etc/ssl/example/ollama.example.com.pem';

    location / {
        proxy_pass                         http://localhost:8080;
        proxy_http_version                 1.1;
        proxy_redirect                     off;
        proxy_set_header Connection        "Upgrade";
        proxy_set_header Host              $host;
        proxy_set_header Upgrade           $http_upgrade;
        proxy_set_header X-Forwarded-For   $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Host  $http_host;
        proxy_set_header X-Forwarded-Port  443;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Real-IP         $remote_addr;
        chunked_transfer_encoding          off;
        proxy_buffering                    off;
        proxy_connect_timeout              3600s;
        proxy_read_timeout                 3600s;
        proxy_send_timeout                 3600s;
        send_timeout                       3600s;
        client_max_body_size               0;
    }
}
```

**Comment by santosh0787** [comment]
hello, anybody able to fix the issue? i have below nginx config 
```
server {
    listen       80;
    server_name  localhost;
    location / {
        root   /usr/share/nginx/html;
        index  index.html index.htm;

    }

    location /chatgpt {
        proxy_pass http://192.168.0.186:3003; 
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }



}


```

Im getting below error in nginx logs
```
2024/11/22 22:36:00 [error] 29#29: *4 open() "/usr/share/nginx/html/_app/immutable/chunks/scheduler.CvZYoJ9P.js" failed (2: No such file or directory), client: 192.168.65.1, server: localhost, request: "GET /_app/immutable/chunks/scheduler.CvZYoJ9P.js HTTP/1.1", host: "localhost", referrer: "http://localhost/chatgpt"
2024/11/22 22:36:00 [error] 28#28: *1 open() "/usr/share/nginx/html/_app/immutable/chunks/index.C1N8d1hw.js" failed (2: No such file or directory), client: 192.168.65.1, server: localhost, request: "GET /_app/immutable/chunks/index.C1N8d1hw.js HTTP/1.1", host: "localhost", referrer: "http://localhost/chatgpt"
2024/11/22 22:36:00 [error] 28#28: *1 open() "/usr/share/nginx/html/static/splash.png" failed (2: No such file or directory), client: 192.168.65.1, server: localhost, request: "GET /static/splash.png HTTP/1.1", host: "localhost", referrer: "http://localhost/chatgpt"
2024/11/22 22:36:00 [error] 28#28: *1 open() "/usr/share/nginx/html/favicon.png" failed (2: No such file or directory), client: 192.168.65.1, server: localhost, request: "GET /favicon.png HTTP/1.1", host: "localhost", referrer: "http://localhost/chatgpt"

```

**Reply by hostingnuggets** [reply]
Is there still no solutions to run Open WebUI behind an nginx reverse proxy under a subdirectory such as `/open-webui` ?

**Comment by spammenotinoz** [comment]
Use a different hostname..:)
I personally have not done it, but use a re-write rule.
rewrite ^/chatgpt(.*)$ $1 break;

**Reply by kitaev-chen** [reply]
already tried that before, it didn't work.

**Reply by hostingnuggets** [reply]
> Use a different hostname..:) I personally have not done it, but use a re-write rule. rewrite ^/chatgpt(.*)$ $1 break;

I also tried that and it does not work.

**Comment by brtptrs** [comment]
Will this commit solve the Problem?
https://github.com/open-webui/open-webui/pull/7415

**Comment by harddal** [comment]
Thought I would add, as of January 2025 I have it working behind nginx (using NPM) by enabling websocket support. Everything works and I don't have any errors. Without websocket support turned on it would still fail to chat.

**Reply by drumtechphoto** [reply]
Enabled Websocket support in NPM hosted on Unraid 6.12.4 and now the chats/prompts are working over the web on my own subdomain.  Thank you @harddal

**Reply by heapsoftware** [reply]
How about using NPM with Websocket and SSL enabled? Once I use HTTPS/SSL its game over.

**Comment by paxprot** [comment]
I think the problem is right now you can't reverse proxy it behind a subdirectory, you have to use the root path

**Comment by ncvescera** [comment]
is there a solution now ?

**Reply by emaadmanzoor** [reply]
+1

**Comment by rostwolke** [comment]
Since I couldn't find anything that worked for me, I checked my previous projects that used WebSockets and Reverse Proxy. Below is my configuration for Apache2 (it needs to be in the vhost):
```
    # Enable Rewrite Engine
    RewriteEngine On

    # 1. Condition: Check if the request is for a WebSocket upgrade
    RewriteCond %{HTTP:Upgrade} =websocket [NC]
    # 2. Rule: If it is, proxy it to the ws:// endpoint
    RewriteRule /(.*)           ws://10.0.15.252:3000/$1 [P,L]

    # 3. All other requests (standard HTTP)
    ProxyPass / http://10.0.15.252:3000/
    ProxyPassReverse / http://10.0.15.252:3000/

    # Optional but recommended security headers (from the post above)
    Header always set X-Frame-Options "deny"
    Header always set X-Content-Type-Options "nosniff"
```
Don't forget to change the URL.

Make sure that the modules `proxy_wstunnel`, `rewrite` and `proxy_http` are enabled:

```
sudo a2enmod proxy_wstunnel
sudo a2enmod rewrite
sudo a2enmod proxy_http
```

I hope this helps others.

**Comment by AbhinayCSC** [comment]
Since ,I am facing issue while using nginx for Open WebUI, below are the details with respect to it:

Nginx code block:

location /fundsgenai/ {

        include headers.conf;

        include timeouts.conf;

        #include restricted.conf;

        proxy_pass http://tgaiocraz01.itgroot.net:3000/;

    }
I am able to access our intranet link http://tgaiocraz01.itgroot.net:3000/fundsgenai/ but while accessing public url https://uat-vu.intertrustgroup.com/fundsgenai/ it is giving an icon when image fails to load.

Can anyone provide suggestions to make it work?
